text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,299 @@
+
+The aim of rehydration is to avoid the client-side performance issues of a naive SSR implementation, reusing original DOM wherever possible.
+
+### The Happy Path
+
+With regard to SSR, there exists a happy path (Fig 1a) where the rendering output for a given request is identical on both the server and the client; identical VDOM is generated for the entire document in both environments.
+
+Depending upon the actual implementation, exceptions to this happy path include:
+
+- **Exception A:** One or more components are never rendered on the server (Fig 1b). Instead, a placeholder is rendered on the server, and later replaced on the client during CSR/hydration.
+  - **Example:** The page contains a cart icon with a dynamic number of items in cart.
+  - **Example:** The page contains a user avatar.
+- **Exception B:** One or more components are rendered only on the server (Fig 1c). No JS is sent to the client for these components and the original SSR'd HTML remains untouched by hydration and later client-side renders.
+  - **Example:** The page contains a footer that never changes.
+  - **Example:** The landing page is mostly static and is cached on a CDN.
+- **Exception C:** Render-critical state is changed between the time of SSR and CSR rehydration, causing an unexpected mismatch (Fig 1d).
+  - **Example:** A product price is updated after SSR but before rehydration on the client.
+  - **Example:** The CDN delivers a stale version of the landing page, which was updated after the SSR HTML was en route to the client.
+
+<figure>
+  <img src="../assets/0000-ssr-rehydration-fig1.png" alt="Fig. 1">
+  <figcaption align = "center">
+    <b>Fig.1: the happy path and its exceptions</b>
+  </figcaption>
+</figure>
+
+## Design
+
+Several pieces need to come together to unlock rehydration. The general flow is illustrated below in figure 2.
+
+<figure>
+  <img src="../assets/0000-ssr-rehydration-fig2.png" alt="Fig. 2" width="60%">
+  <figcaption align = "center">
+    <b>Fig.2: the proposed implementation</b>
+  </figcaption>
+</figure><br /><br />
+
+In roughly chronological order with respect to a single page request, the following steps shall be taken as part of rehydration:
+
+- As part of SSR, serialize render-critical state and bake it into the HTML response returned from the server. For each `LightningElement`, attach corresponding state to its root DOM node as `data-ssr-state`.
+
+- Rather than calling `LWC#createElement`, required that `LWC#hydrateElement` is called to mount the root Lightning Web Component on the client.
+
+- Generate server-side VDOM from the SSR HTML on the client.
+
+- As the initial client-side render progresses, extract state from `data-ssr-state` attributes, instantiate `LightningElements` with extracted state, and generate client-side VDOM.

```

Comment:
My thought is that it could be present on any component in the subtree. The other approach with the same aim is to extract a state tree that mirrors the component tree during SSR, and serialize state in a single place. However, serializing the state at the root node of each component means the state is easily accessible during traversal in rehydration, which is attractive from the implementation side. Open to feedback here, if you think there are pitfalls I haven't accounted for.